core/src/main/scala/scalaz/ImmutableArray.scala

```diff
@@ -15,15 +15,16 @@ import syntax.Ops
 sealed abstract class ImmutableArray[+A] {
   protected[this] def elemTag: ClassTag[A]
 
-  def apply(index: Int): A
+  // these methods are not total
+  private[scalaz] def apply(index: Int): A
+  private[scalaz] def copyToArray[B >: A](xs: Array[B], start: Int, len: Int): Unit
+  private[scalaz] def slice(from: Int, until: Int): ImmutableArray[A]
```

Comment:
I think slice is total.

Reply:
`slice(-1, -1)` ?

Reply:
`slice(-1, -1)` return empty ImmutableArray. not throw Exception

Reply:
ok, so it basically silently fails with a poison pill. Are we sure we want that behaviour?

Reply:
I think there are some such methods in scalaz from one point of view. e.g. `IList#slice`, `IList#drop`, `IList#take`, `IList#takeRight`, `ISet#delete` ...

Reply:
ok, updated for now. `toArray` is probably not a good method to have since it is returning a mutable object. Should we make it private?